src/m365/spo/commands/listitem/listitem-retentionlabel-ensure.ts

```diff
@@ -0,0 +1,198 @@
+          !validation.isValidGuid(args.options.id as string)) {
+          return `${args.options.id} is not a valid GUID`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push({ options: ['listId', 'listTitle', 'listUrl'] });
+    this.optionSets.push({ options: ['name', 'id'] });
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    try {
+      const labelInformation = await this.getLabelInformation(args.options, logger);
+      await this.applyLabel(args.options, labelInformation, logger);
+    }
+    catch (err: any) {
+      this.handleRejectedODataJsonPromise(err);
+    }
+  }
+
+  private async getLabelInformation(options: Options, logger: Logger): Promise<any> {
+    const cmdOptions: SpoWebRetentionLabelListCommandOptions = {
+      webUrl: options.webUrl,
+      output: 'json',
+      debug: options.debug,
+      verbose: options.verbose
+    };
+
+    const output = await Cli.executeCommandWithOutput(SpoWebRetentionLabelListCommand as Command, { options: { ...cmdOptions, _: [] } });
+
+    if (this.verbose) {
+      logger.logToStderr(output.stderr);
+    }
+
+    const labels = JSON.parse(output.stdout) as SiteRetentionLabel[];
+    const label = labels.find(l => l.TagName === options.name || l.TagId === options.id);
+
+    if (this.verbose && label !== undefined) {
+      logger.logToStderr(`Retention label found in the list of available labels: '${label.TagName}' / '${label.TagId}'...`);
+    }
+
+    if (label === undefined) {
+      throw new Error(`The specified retention label does not exist`);
+    }
+
+    const requestBody = {
+      "complianceTag": label.TagName,
+      "isTagPolicyHold": label.BlockDelete,
+      "isTagPolicyRecord": label.BlockEdit,
+      "isEventBasedTag": label.IsEventTag,
+      "isTagSuperLock": label.SuperLock,
+      "isUnlockedAsDefault": label.UnlockedAsDefault
+    };
+
+    return requestBody;
+  }
+
+  private async applyLabel(options: Options, labelInformation: any, logger: Logger): Promise<void> {
+    if (this.verbose) {
+      logger.logToStderr(`Applying retention label to item in list '${options.listId || options.listTitle || options.listUrl}' in site at ${options.webUrl}...`);
+    }
+
+    let requestUrl = `${options.webUrl}/_api/web`;
+
+    if (options.listId) {
+      requestUrl += `/lists(guid'${formatting.encodeQueryParameter(options.listId)}')/items(${options.listItemId})/SetComplianceTag()`;
+    }
+    else if (options.listTitle) {
+      requestUrl += `/lists/getByTitle('${formatting.encodeQueryParameter(options.listTitle)}')/items(${options.listItemId})/SetComplianceTag()`;
+    }
+    else if (options.listUrl) {
+      const listServerRelativeUrl: string = urlUtil.getServerRelativePath(options.webUrl, options.listUrl);
+      requestUrl += `/GetList(@a1)/items(${options.listItemId})/SetComplianceTag()?@a1='${formatting.encodeQueryParameter(listServerRelativeUrl)}'`;
```

Comment:
why not just? 
```suggestion
      requestUrl += `/GetList('${formatting.encodeQueryParameter(listServerRelativeUrl)}')/items(${options.listItemId})/SetComplianceTag()`;
```

Reply:
That's also possible. The road I pick here is on purpose, as there is a small chance that the entire URL is longer than the URL limit. You can circumvent that by using this setup with a query parameter.

Reply:
ok clear 👍. In that case I would suggest to use meaningful query parameter naming like `@listUrl` instead of `@a` 😉

Reply:
ok, fair point.